### PR TITLE
Adding Ring and Belt Dynamics

### PR DESCRIPTION
**Pull request by NoFoolLikeOne**
Uses a nominal radius = Inner +( (outer-inner)*(3/8)) and then applies Kepler's Third Law

image

### CODE REVIEW

File: src/app/system-body/system-body.component.ts [hunk 1]
```diff
@@ -380,6 +386,45 @@ export class SystemBodyComponent implements OnChanges {
...
+      const earthMassKg = 5.972e24;
+      const parentMassKg = parentBd.solarMasses != null
+        ? parentBd.solarMasses * solarMassKg
+        : (parentBd.earthMasses != null ? parentBd.earthMasses * earthMassKg : 0);
+      if (parentMassKg > 0) {
+        const nominalRadiusM = (inner + (outer - inner) * (3 / 8)) * 1000;
+        const outerRadiusM = outer * 1000;
+        const periodS = 2 * Math.PI * Math.sqrt(Math.pow(nominalRadiusM, 3) / (G * parentMassKg));
+        const periodDays = periodS / 86400;
+        const maxVelocityKms = (2 * Math.PI * outerRadiusM / periodS) / 1000;
```
Comment (DarkSession) [inline src/app/system-body/system-body.component.ts]: **'Max velocity' mixes a Keplerian period with a rigid-disc velocity model.** The period above is the Keplerian orbital period at the *nominal* radius, but here that single period is applied to the *outer* circumference (`2π·r_outer/T`) — i.e. a rigid disc where every radius shares one angular velocity. Real ring/belt particles orbit independently at `v = √(GM/r)`, so speed is highest at the **inner** edge and lowest at the outer edge. For Saturn's rings this reports ~29.7 km/s, whereas the true maximum (inner edge) is ~23.8 km/s and the outer edge is only ~16.5 km/s — so the value is ~25% high and at the wrong edge. Either compute the true inner-edge maximum `√(GM/r_inner)`, or rename the row to make the rigid-rotation assumption explicit.

File: src/app/system-body/system-body.component.ts [hunk 2]
```diff
@@ -380,6 +386,45 @@ export class SystemBodyComponent implements OnChanges {
...
+    if (outer > 0 && inner >= 0 && body.parent) {
+      const parentBd = body.parent.bodyData;
+      const G = 6.674e-11; // m³/(kg·s²)
+      const solarMassKg = 1.989e30;
+      const earthMassKg = 5.972e24;
+      const parentMassKg = parentBd.solarMasses != null
+        ? parentBd.solarMasses * solarMassKg
+        : (parentBd.earthMasses != null ? parentBd.earthMasses * earthMassKg : 0);
+      if (parentMassKg > 0) {
+        const nominalRadiusM = (inner + (outer - inner) * (3 / 8)) * 1000;
```
Comment (DarkSession) [inline src/app/system-body/system-body.component.ts]: **Magic number `3/8` for the nominal radius.** This isn't the annulus area centroid (`(2/3)(R_o³−R_i³)/(R_o²−R_i²)`, ≈0.56 of inner→outer for Saturn's rings) nor the simple mid-radius. Worth a comment justifying the choice, or use a principled representative radius.

File: src/app/system-body/system-body.component.ts [hunk 3]
```diff
@@ -380,6 +386,45 @@ export class SystemBodyComponent implements OnChanges {
     this.isRingNotVisible.set(bd.type === BODY_TYPE.Ring
       && this.getRingDensity() < 0.1 && this.getRingWidth() > 1000000);
 
+    // Ring dynamics: orbital period and max velocity.
+    if (outer > 0 && inner >= 0 && body.parent) {
```
Comment (DarkSession) [inline src/app/system-body/system-body.component.ts]: **`inner >= 0` is a no-op guard.** `inner` is `bd.innerRadius ?? 0`, so it is always ≥ 0 and can never exclude anything — only `outer > 0` does real filtering. Drop it, or replace it with the condition actually intended.

File: src/app/system-body/system-body.component.ts [hunk 4]
```diff
@@ -380,6 +386,45 @@ export class SystemBodyComponent implements OnChanges {
...
+      const parentBd = body.parent.bodyData;
+      const G = 6.674e-11; // m³/(kg·s²)
+      const solarMassKg = 1.989e30;
```
Comment (DarkSession) [inline src/app/system-body/system-body.component.ts]: **Duplicated physical constants.** `solarMassKg`/`earthMassKg` (and `G`) duplicate `KG_PER_SOLAR_MASS`/`KG_PER_EARTH_MASS` already defined in `body-physics.service.ts`, which also exposes a `parentMassEarthMasses()` helper doing this exact solar→earth fallback. Reuse those rather than redefining the constants here.

**Review by DarkSession — COMMENTED** [review]
Automated review of the new Ring/Belt Dynamics feature — main points are inline.

One note that isn't tied to a single line: **the new logic has no unit tests.** No spec references `getRingRotationalPeriod`, `getRingMaxVelocity`, or `formatVelocityKms`. The e2e `Dynamics → Rotational period` assertions cover the pre-existing *body* Dynamics section, not this ring/belt one. Given the project's ≥80% coverage expectation, please add tests for the period formula, the parent-mass fallback, the hidden/null paths, and the km/s↔c threshold.

File: src/app/system-body/system-body.component.ts [hunk 5]
```diff
@@ -380,6 +386,45 @@ export class SystemBodyComponent implements OnChanges {
     this.isRingNotVisible.set(bd.type === BODY_TYPE.Ring
       && this.getRingDensity() < 0.1 && this.getRingWidth() > 1000000);
 
+    // Ring dynamics: orbital period and max velocity.
```
Comment (DarkSession) [inline src/app/system-body/system-body.component.ts]: **This physics arguably belongs in `body-physics.service.ts`.** The codebase centralizes ring/orbital physics in that service (Roche/Hill limits, density — all delegated just a few lines below). Inlining ~40 lines of Kepler math in the component breaks that pattern and isn't independently unit-testable. Consider moving it into the service alongside the other ring calculations.

File: src/app/system-body/system-body.component.ts [hunk 6]
```diff
@@ -380,6 +386,45 @@ export class SystemBodyComponent implements OnChanges {
...
+    if (outer > 0 && inner >= 0 && body.parent) {
+      const parentBd = body.parent.bodyData;
+      const G = 6.674e-11; // m³/(kg·s²)
+      const solarMassKg = 1.989e30;
+      const earthMassKg = 5.972e24;
+      const parentMassKg = parentBd.solarMasses != null
+        ? parentBd.solarMasses * solarMassKg
+        : (parentBd.earthMasses != null ? parentBd.earthMasses * earthMassKg : 0);
+      if (parentMassKg > 0) {
```
Comment (DarkSession) [inline src/app/system-body/system-body.component.ts]: **Silent suppression when the parent has no mass.** When the parent exposes neither `solarMasses` nor `earthMasses`, `parentMassKg` is 0 and the whole Dynamics section disappears with no indication to the user. Consider surfacing a 'data unavailable' state rather than hiding it silently.

File: src/app/system-body/system-body.component.ts [hunk 7]
```diff
@@ -380,6 +386,45 @@ export class SystemBodyComponent implements OnChanges {
...
+        this.getRingMaxVelocityTooltip.set(`${maxVelocityKms.toFixed(3)} km/s`);
+      } else {
+        this.getRingRotationalPeriod.set(null);
+        this.getRingRotationalPeriodDisplay.set('');
+        this.getRingRotationalPeriodTooltip.set('');
+        this.getRingMaxVelocity.set(null);
+        this.getRingMaxVelocityDisplay.set('');
+        this.getRingMaxVelocityTooltip.set('');
+      }
+    } else {
```
Comment (DarkSession) [inline src/app/system-body/system-body.component.ts:405]: **Duplicated reset block.** These six resets are identical to the `else` branch just above (lines 411-418). Initialise all six signals to `null`/`''` once before the `if`, then only set them on the success path — that removes both copies.

File: src/app/system-body/system-body.component.ts [hunk 8]
```diff
@@ -1069,6 +1114,15 @@ export class SystemBodyComponent implements OnChanges {
         body.bodyData.subType === 'Neutron Star');
   }
 
+  private formatVelocityKms(velocityKms: number): string {
+    const speedOfLight = 299792.458; // km/s
+    const fractionOfC = velocityKms / speedOfLight;
+    if (fractionOfC >= 0.01) {
```
Comment (DarkSession) [inline src/app/system-body/system-body.component.ts:1218]: **Effectively-dead `c` branch.** This switches to fraction-of-c units at ≥0.01c (~2998 km/s); ring/belt speeds are tens of km/s, so it never triggers for realistic data (and if it did, it would be formatting the rigid-rotation value flagged on the velocity calc). Low priority — fine to keep as defensive, but note it's currently unreachable.

File: src/app/system-body/system-body.component.html [hunk 9]
```diff
@@ -522,6 +522,32 @@
...
             </div>
           }
+          <!-- Dynamics Section (Rings and Belts) -->
+          @if ((body().bodyData.type === 'Ring' || body().bodyData.type === 'Belt') && (getRingRotationalPeriod() !== null || getRingMaxVelocity() !== null)) {
+            <div class="body-data-section">
+              <div class="body-data-section-header">Dynamics</div>
+              @if (getRingRotationalPeriod() !== null) {
+                <div class="body-data-entry">
+                  <div [matTooltip]="'Rotational period calculated at nominal radius (inner + (outer−inner)×3/8)'">
+                    Rotational period
```
Comment (DarkSession) [inline src/app/system-body/system-body.component.html]: **'Rotational period' is a slight misnomer.** This is the *orbital* period of a particle at the nominal radius; particles at other radii have different periods. Presenting one radius's period as the ring's single 'Rotational period' carries the same rigid-disc simplification as 'Max velocity' — worth making the assumption explicit in the label or tooltip.

File: src/app/system-body/system-body.component.html [hunk 10]
```diff
@@ -522,6 +522,32 @@
               }
             </div>
           }
+          <!-- Dynamics Section (Rings and Belts) -->
+          @if ((body().bodyData.type === 'Ring' || body().bodyData.type === 'Belt') && (getRingRotationalPeriod() !== null || getRingMaxVelocity() !== null)) {
+            <div class="body-data-section">
+              <div class="body-data-section-header">Dynamics</div>
+              @if (getRingRotationalPeriod() !== null) {
```
Comment (DarkSession) [inline src/app/system-body/system-body.component.html]: **Redundant inner `@if`s.** `getRingRotationalPeriod()` and `getRingMaxVelocity()` are always set together (both null or both non-null), so this inner guard and the one for Max velocity below are effectively always true whenever the outer `@if` passes. Minor redundancy.

File: src/app/system-body/system-body.component.ts (hunk 1 above)
Comment (NoFoolLikeOne) [reply src/app/system-body/system-body.component.ts]: In Elite Dangerous all parts of the ring orbit with the same period. So the tangential velocity of the inner ring is slower than the outer rings. Real world physics does not apply.

File: src/app/system-body/system-body.component.ts (hunk 2 above)
Comment (NoFoolLikeOne) [reply src/app/system-body/system-body.component.ts]: The 3/8 figure has been arrived at using observational data. 
We orginally though e (Euler's number) was what we could use to find the observational nominal radius. But the closest fit to our data we 3/8 which is about midway between e and 1/φ² 

We are still gathering data but it is helpful to have the measurement in the tool as a good approximation. It is not possible to get 100% accuracy because the game journals store the ring radii in sci4d so there is not enough precision and we may yet find exceptional data that proves it wrong.

File: src/app/system-body/system-body.component.ts (hunk 5 above)
Comment (NoFoolLikeOne) [reply src/app/system-body/system-body.component.ts]: Guess my Claude didn't think of that

File: src/app/system-body/system-body.component.ts (hunk 6 above)
Comment (NoFoolLikeOne) [reply src/app/system-body/system-body.component.ts]: If there is no mass or solarMass then I think leaving this section out is fine given that Mass is separately displayed piece of data which should indicate data is unavailable

File: src/app/system-body/system-body.component.ts (hunk 7 above)
Comment (NoFoolLikeOne) [reply src/app/system-body/system-body.component.ts:405]: Sounds good

File: src/app/system-body/system-body.component.ts (hunk 8 above)
Comment (NoFoolLikeOne) [reply src/app/system-body/system-body.component.ts:1218]: Its not dead Flyae Flyuae AA-A h35 AB 10 A Ring is a fraction of c and I think it is appropriate to display it as such because it highlights the extreme nature of the system

File: src/app/system-body/system-body.component.html (hunk 9 above)
Comment (NoFoolLikeOne) [reply src/app/system-body/system-body.component.html]: In Elite dangerous particles all have the same period, So effectively it is rigid disk. 
Making this explicit is a bit tricky because it highlights that the game is not scientifically accurate. Could take the same approach and point out that 21st Century Science got it wrong

File: src/app/system-body/system-body.component.ts [hunk 11]
```diff
@@ -1087,6 +1096,33 @@ export class SystemBodyComponent implements OnChanges {
...
+      this.getRingMaxVelocity.set(null);
+      this.getRingMaxVelocityDisplay.set('');
+      this.getRingMaxVelocityTooltip.set('');
+    }
+  }
+
+  private formatVelocityKms(velocityKms: number): string {
+    const speedOfLight = 299792.458; // km/s
+    const fractionOfC = velocityKms / speedOfLight;
+    if (fractionOfC >= 0.01) {
```
Comment (DarkSession) [inline src/app/system-body/system-body.component.ts:1218]: **No test for the km/s↔c threshold.** `formatVelocityKms` is private and reached only via `getRingMaxVelocityDisplay`; no spec produces a velocity ≥ 0.01c, so this `c` branch — the headline "extreme system" case (e.g. Flyae Flyuae AA-A h35 AB 10 A Ring) — is untested. The `ringDynamics` service tests cover the maths but not this formatting. Worth a small test asserting the `…c` vs `… km/s` switch (and ideally `formatPeriodDays` for rings).

File: src/app/system-body/system-body.component.ts [hunk 12]
```diff
@@ -1087,6 +1096,33 @@ export class SystemBodyComponent implements OnChanges {
...
+      this.getRingOrbitalPeriodDisplay.set('');
+      this.getRingOrbitalPeriodTooltip.set('');
+      this.getRingMaxVelocity.set(null);
+      this.getRingMaxVelocityDisplay.set('');
+      this.getRingMaxVelocityTooltip.set('');
+    }
+  }
+
+  private formatVelocityKms(velocityKms: number): string {
+    const speedOfLight = 299792.458; // km/s
```
Comment (DarkSession) [inline src/app/system-body/system-body.component.ts]: **Duplicated speed-of-light constant.** This hardcodes `299792.458` (km/s) and line 1067 hardcodes `299792458` (m/s), while `body-physics.service.ts` already defines `SPEED_OF_LIGHT = 299792458` (currently un-exported). Same class of duplication that was flagged earlier for the mass constants — consider exporting `SPEED_OF_LIGHT` from the service and reusing it in both places.

File: src/app/data/body-physics.service.ts [hunk 13]
```diff
@@ -246,6 +260,50 @@ export class BodyPhysicsService {
...
+   *  - The game journals store ring radii to 4 significant figures, limiting precision.
+   *  - Data collection is ongoing; edge cases may yet refine or challenge the factor.
+   *
+   * The maximum velocity is the tangential speed at the outer edge using the period
+   * derived from the nominal radius.
+   *
+   * @param body  The Ring or Belt SystemBody.
+   * @returns     Dynamics values, or null when parent mass or radii are unavailable.
+   */
+  ringDynamics(body: SystemBody): RingDynamics | null {
```
Comment (DarkSession) [inline src/app/data/body-physics.service.ts:307]: **`ringDynamics` isn't gated by body type.** It returns values for any body with `outerRadius > 0` and a parent; the Ring/Belt check lives only in the template. Harmless in practice (only rings/belts carry `outerRadius`), but the method name implies a contract it doesn't enforce. An early `if (bd.type !== BODY_TYPE.Ring && bd.type !== BODY_TYPE.Belt) { return null; }` would make it explicit and keep the service self-contained / independently testable.

File: src/app/system-body/system-body.component.html [hunk 14]
```diff
@@ -628,27 +556,54 @@
           }
         </div>
         }
+        <!-- Dynamics Section (Rings and Belts) -->
+        @if ((body().bodyData.type === 'Ring' || body().bodyData.type === 'Belt') && (getRingOrbitalPeriod() !== null
+        || getRingMaxVelocity() !== null)) {
+        <div class="body-data-section">
+          <div class="body-data-section-header">Dynamics</div>
+          @if (getRingOrbitalPeriod() !== null) {
```
Comment (DarkSession) [inline src/app/system-body/system-body.component.html:570]: **Redundant inner `@if`s.** `applyRingDynamics` always sets `getRingOrbitalPeriod` and `getRingMaxVelocity` together (both non-null or both null), and the outer `@if` already checks `period || velocity`. So this guard and the Max-velocity one below are always true once the outer `@if` passes. Minor — can be dropped.

File: src/app/data/body-physics.service.spec.ts [hunk 15]
```diff
@@ -108,6 +108,91 @@ describe('BodyPhysicsService', () => {
...
+    // The 3/8 factor is empirically derived from observational data.
+
+    const G = 6.6743e-11;
+    const KG_PER_EARTH_MASS = 5.972e24;
+    // Mirror the service's conversion: solarMasses → earthMasses → kg
+    // (uses EARTH_MASSES_PER_SOLAR_MASS = 332950, not a direct KG_PER_SOLAR_MASS constant).
+    const EARTH_MASSES_PER_SOLAR_MASS = 332950;
+    const KG_PER_SOLAR_MASS_VIA_EARTH = EARTH_MASSES_PER_SOLAR_MASS * KG_PER_EARTH_MASS;
+
+    function expectedDynamics(innerKm: number, outerKm: number, massKg: number) {
```
Comment (DarkSession) [inline src/app/data/body-physics.service.spec.ts:129]: **Self-referential expectations.** `expectedDynamics` mirrors the service formula exactly (same `×3/8`, same `×1000`), so these tests confirm internal consistency but would still pass if a unit conversion were wrong. Consider pinning at least one hard-coded reference value — e.g. inner=0, outer=8000 km around a 1 Earth-mass parent → ≈ 0.0189272 days and ≈ 30.738 km/s — so the formula is anchored to an independent number.

**Review by DarkSession — COMMENTED** [review]
Follow-up review of the Ring/Belt Dynamics code after the post-review commits (physics moved into `BodyPhysicsService.ringDynamics`, physical constants reused, the `3/8` factor documented, the no-op `inner >= 0` guard removed, the label changed to 'Orbital period', and service-level tests added). Those resolve the bulk of the earlier review — nice work.

I re-checked the one correctness concern from the first pass: the `×1000` km→m conversion in `ringDynamics` is **correct**. The standalone Ring/Belt body's `bodyData.outerRadius`/`innerRadius` are in **km** (displayed directly as `… km`), which is distinct from the metres-based `parent.rings[]` array (divided by 1000 elsewhere). So there's no units bug.

A few smaller, non-blocking items remain (inline). All were independently verified against HEAD (`9a54fc9`).

File: src/app/system-body/system-body.component.html (hunk 14 above)
Comment (DarkSession) [reply src/app/system-body/system-body.component.html:570]: Withdrawing this. The outer `@if` uses OR (`period !== null || velocity !== null`), so the inner guards are the correct complement — they decouple the two rows and render each only when its own value is present. They're only "always true" because `applyRingDynamics` currently sets both signals together (a .ts invariant the template can't see); if the two ever diverged, these guards would correctly avoid rendering an empty row. Defensive and intentional, not redundant.

File: src/app/system-body/system-body.component.ts [hunk 16]
```diff
@@ -1087,6 +1129,73 @@ export class SystemBodyComponent implements OnChanges {
...
   }
 
+  private isInvisibleRing(bd: CanonnBiostatsBody): boolean {
+    if (bd.type !== BODY_TYPE.Ring) { return false; }
+    const outer = bd.outerRadius ?? 0;
+    const inner = bd.innerRadius ?? 0;
+    const width = outer - inner;
+    const area = Math.PI * (outer * outer - inner * inner);
+    const density = area > 0 ? (bd.mass ?? 0) / area : 0;
+    return density < 0.1 && width > 1000000;
```
Comment (DarkSession) [inline src/app/system-body/system-body.component.ts]: **Triplicated invisible-ring logic / magic numbers.** The predicate `density < 0.1 && width > 1000000` (recomputing width/area/density each time) now appears three times: here in `isInvisibleRing`, in the `isRingNotVisible` signal (~line 359), and in `showInvisibleRingExplanation` (~line 838). This `isInvisibleRing` also recomputes the values instead of reusing the cached `getRingWidth()`/`getRingArea()`/`getRingDensity()` signals. Same duplication class previously flagged for the mass/speed-of-light constants — suggest a single shared predicate with named threshold constants.

File: src/app/system-body/system-body.component.ts [hunk 17]
```diff
@@ -1087,6 +1129,73 @@ export class SystemBodyComponent implements OnChanges {
...
+  }
+
+  private computeRingNeighbourDistance(body: SystemBody): { distance: number | null; label: string; velocityDiff: number | null; eitherRingInvisible: boolean } {
+    const bd = body.bodyData;
+    if ((bd.type !== BODY_TYPE.Ring && bd.type !== BODY_TYPE.Belt) || !body.parent) {
+      return { distance: null, label: '', velocityDiff: null, eitherRingInvisible: false };
+    }
+    const siblings = body.parent.subBodies
+      .filter(s => s.bodyData.name.includes('Ring'))
```
Comment (DarkSession) [inline src/app/system-body/system-body.component.ts:1173]: **Dead `Belt` branch.** `computeRingNeighbourDistance` early-returns unless the body is a Ring *or* Belt, but this sibling filter keys on `name.includes('Ring')`. After `stripParentName`, belt names are `"A Belt"` (etc.) and never contain `"Ring"`, so a Belt always gets `idx = -1` and returns null — the `|| type === Belt` part of the gate is unreachable. Either drop Belt from the gate, or include belts in the sibling filter if neighbour-distance is meant to apply to them.

File: src/app/system-body/system-body.component.ts [hunk 18]
```diff
@@ -343,6 +359,33 @@ export class SystemBodyComponent implements OnChanges {
...
+    this.getRingNeighbourDistance.set(neighbourDist);
+    this.getRingNeighbourDistanceLabel.set(neighbourLabel);
+    this.getRingVelocityDiff.set(velocityDiff);
+    this.getRingVelocityDiffDisplay.set(velocityDiff !== null ? this.formatVelocityKms(velocityDiff) : '');
+    this.getRingVelocityDiffTooltip.set(velocityDiff !== null ? `${velocityDiff.toFixed(3)} km/s` : '');
+
+    // Racing Rings badge: inner ring gap < 50 km AND velocity difference > 5 km/s,
+    // and neither ring in the pair is invisible.
+    const racingRings = neighbourDist !== null && velocityDiff !== null
+      && neighbourDist < 50 && velocityDiff > 5 && !eitherRingInvisible;
```
Comment (DarkSession) [inline src/app/system-body/system-body.component.ts]: **Undocumented Racing-Rings thresholds.** The `< 50` km gap and `> 5` km/s differential are bare literals. Given the `3/8` nominal-radius factor got a full justification block, these heuristic cutoffs deserve named constants or a short note on their empirical basis.

File: src/app/system-body/system-body.component.ts [hunk 19]
```diff
@@ -1087,6 +1129,73 @@ export class SystemBodyComponent implements OnChanges {
     return this.stellarPhysics.radiusKm(bd.radius, bd.solarRadius);
   }
 
+  private isInvisibleRing(bd: CanonnBiostatsBody): boolean {
```
Comment (DarkSession) [inline src/app/system-body/system-body.component.ts:1152]: **Build is broken — `CanonnBiostatsBody` is not imported.** `ng build` fails here:

```
✘ [ERROR] TS2304: Cannot find name 'CanonnBiostatsBody'.
    src/app/system-body/system-body.component.ts:1132:30:
      1132 │   private isInvisibleRing(bd: CanonnBiostatsBody): boolean {
```

`isInvisibleRing` takes a `CanonnBiostatsBody`, but the import on line 3 only pulls in `SystemBody, EdGalaxyData`. Add `CanonnBiostatsBody` (exported from `../home/home.component`) to that import:

```ts
import { SystemBody, EdGalaxyData, CanonnBiostatsBody } from '../home/home.component';
```

Comment (NoFoolLikeOne) [reply src/app/system-body/system-body.component.ts:1152]: I'm not seeing this one. I tested the build before pushing the changes,

**Review by DarkSession — APPROVED** [review]
